### ISSUE TITLE
Can I know whether a decoded value is a `Multiaddress` or a `Balance`?

### ISSUE THREAD

**Issue by Tbaut** [polkadot-api/polkadot-api#762]
This is probably more a feature request than a real question, unless I missed something.

I'm displaying generically some decoded calls for users to review (I guess any wallet would need this). I'd like to display addresses nicely, or balances with the right decimals. Is there a way to detect whether some value is of a certain type.

I can't really rely on checking whether the type is `bigint` or guessing if the value is an address. I tried to check the `instanceof` for an  address, but it seems that it's decoded as type `string` simply.

FWIW with pjs, we had a property `typeName` that was far from perfect, since we had to manually list all the possible names, but did the job so far.
```ts
['Balance', 'BalanceOf', 'Amount'].includes(typeName)  // it's a balance, display with the right decimals
['AccountIdLookupOf', 'LookupSource'].includes(typeName) // it's an address, display with the identicon
```

**Comment by Tbaut** [comment]
Any feedback on this would be greatly appreciated, whether this makes sense to you or not :pray:

**Comment by voliva** [comment]
> I'm displaying generically some decoded calls for users to review (I guess any wallet would need this). I'd like to display addresses nicely, or balances with the right decimals. Is there a way to detect whether some value is of a certain type.

How are you decoding this data?

If you are using the the top-level client, it decodes it directly to the actual value, so there's little we can do there, as we can't add "metadata" to primitive values like bigints...

As for lower-level tooling, we have the @polkadot-api/view-builder, which I think it's main use case is this: it decodes a call data, but giving intermediate values.

Unfortunately we lack documentation on this package, so this issue is a nice reminder to give it some extra love. Please, take a look by yourself, the snapshot tests we have can probably help you see what it does, and if it could what you're looking for. Try console logging the result of the [first test](https://github.com/polkadot-api/polkadot-api/blob/main/packages/view-builder/tests/view-builder.spec.ts#L13) to explore what you can find inside.

**Comment by Tbaut** [comment]
> How are you decoding this data?

Yes, I decode it to display it for now, but it's not what users are happy with since it's too barebone. I'm really only looking to render nicely a couple things such as the balance which is an absolute must to show to users nicely, or the accounts to show the identicon, or some on-chain identity etc.

Thanks for the pointer, the [snapshot](https://github.com/polkadot-api/polkadot-api/blob/main/packages/view-builder/tests/__snapshots__/view-builder.spec.ts.snap) from the test you linked is indeed very interesting. I could eventually use a combination of the `view-builder` and the decoded data for convenience, but that sounds hacky, or I would take a decoder you have and add some metadata fields to clearly identity certain types. I wonder how you do @tien to identfy certain fields like accounts in the extrinsic builder.  This is a common need for any wallet, so it'd be good to unite and make the best out of it. I know @josepot was working on something to help UIs to create extrinsincs as well, this is also a need I have for Multix with a generic extrinsic builder beyond the generic extrinsic display.

**Comment by tien** [comment]
Hey @Tbaut,

Long answer: I have to use a combination of the `dynamicBuilder` + `viewBuilder`, which can be referenced [here](https://github.com/tien/dot-console/blob/main/src/routes/_layout/extrinsics.tsx).

Short answer: Might be best to just wait on the work that @josepot is working on, so that you don't have to go through all the trouble.
